src/ui/SearchInterface/ScrollRestoration.ts

```diff
@@ -0,0 +1,89 @@
+import { QueryStateModel } from '../../models/QueryStateModel';
+
+export interface IScrollRestorationInfo {
+  lastPosition: number;
+  pageHeight: number;
+}
+
+export class ScrollRestoration {
+  static ID = 'ScrollRestoration';
+
+  private scrollInfoStorage: StorageUtils<Record<string, IScrollRestorationInfo>>;
+
+  private restorationTimeOutInMs: number = 5000;
+  private tryToScrollIntervalInMs: number = 50;
+  private timeoutHandle = null;
+
+  constructor(public element: HTMLElement, private queryStateModel: QueryStateModel) {
+    this.scrollInfoStorage = new StorageUtils<Record<string, IScrollRestorationInfo>>(ScrollRestoration.ID, 'session');
+    window.addEventListener('beforeunload', () => {
+      this.saveScrollPositionInState();
+    });
+    $$(this.element).on(ResultListEvents.newResultsDisplayed, (event: Event, args: IDisplayedNewResultsEventArgs) =>
+      this.handleNewResultsDisplayed(args)
+    );
+  }
+
+  public saveScrollPositionInState() {
+    const scrollInfo: IScrollRestorationInfo = {
+      pageHeight: window.document.body.scrollHeight,
+      lastPosition: window.pageYOffset
+    };
+
+    const key = this.getKeyForCurrentQuery();
+    const prevScrollInfo = this.scrollInfoStorage.load();
+    this.scrollInfoStorage.save({
+      ...prevScrollInfo,
+      [key]: scrollInfo
+    });
+  }
+
+  public handleNewResultsDisplayed(args: IDisplayedNewResultsEventArgs) {
+    if (args.infiniteScrolling) {
+      new Logger(this).warn('Scroll restoration is not yet supported on result lists with infinite scrolling enabled.');
+      this.resetScrollPositionState();
+      return;
+    }
+
+    const scrollInfo = this.getScrollInfoForCurrentQuery();
+
+    this.resetScrollPositionState();
+
+    if (scrollInfo && scrollInfo.lastPosition) {
+      const stopTryingAt = Date.now() + this.restorationTimeOutInMs;
+      this.tryToRestoreScrollPosition(scrollInfo, stopTryingAt);
+    }
+  }
+
+  private resetScrollPositionState() {
+    this.scrollInfoStorage.remove(this.getKeyForCurrentQuery());
+  }
+
+  private tryToRestoreScrollPosition(scrollInfo: IScrollRestorationInfo, stopAt: number) {
+    const html = window.document.documentElement;
+    const body = window.document.body;
+
+    const documentHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);
+
+    clearTimeout(this.timeoutHandle);
+
+    if (documentHeight >= scrollInfo.pageHeight || Date.now() >= stopAt) {
+      window.scrollTo(0, scrollInfo.lastPosition);
+    } else {
+      this.timeoutHandle = setTimeout(() => this.tryToRestoreScrollPosition(scrollInfo, stopAt), this.tryToScrollIntervalInMs);
+    }
+  }
+
+  private getScrollInfoForCurrentQuery() {
+    const scrollInfo = this.scrollInfoStorage.load() || {};
+    return scrollInfo[this.getKeyForCurrentQuery()];
+  }
+
+  private getKeyForCurrentQuery() {
+    return JSON.stringify(this.queryStateModel.getAttributes());
```

Comment:
I "think" that's okay 🤔 
Wondering if there's some kind of limit to the key length for session storage ? 
If yes, it's possible that it could get "too large" and fail in a random/hard to debug way ?

Should we have some kind of `hashCode()` function ?

Reply:
In fact, there's only one session storage key called 'Coveo-ScrollRestoration', and it holds a map-like object with the query parameters as key. Supposedly, browsers can handle at least strings and property names up to at least 2^27 (https://stackoverflow.com/questions/13367391/is-there-a-limit-on-length-of-the-key-string-in-js-object). But it doesn't hurt to try the hashcode approach. `StringUtils.hashCode()` could do it right?

Reply:
Well if it's not needed because no bugs could occur, then I would not needlessly add complexity to the code.